src/main/java/edu/wpi/grip/core/sources/WebCamSource.java

```diff
@@ -0,0 +1,131 @@
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Provides a way to generate a constantly updated {@link Mat} from a WebCam attached
+ * to the users computer.
+ */
+public class WebCamSource implements Source {
+
+    private final SocketHint<Mat> imageOutputHint = new SocketHint<Mat>("Image", Mat.class, Mat::new);
+    private final SocketHint<Number> framerateOutputHint = new SocketHint<Number>("Framerate", Number.class, 0);
+    private OutputSocket<Mat> frameOutputSocket;
+    private OutputSocket<Number> frameRateOutputSocket;
+    private Optional<Thread> executor;
+    private Optional<OpenCVFrameGrabber> grabber;
+
+    public WebCamSource(EventBus eventBus) {
+        checkNotNull(eventBus, "Event Bus was null.");
+        frameOutputSocket = new OutputSocket(eventBus, imageOutputHint);
+        frameRateOutputSocket = new OutputSocket(eventBus, framerateOutputHint);
+        grabber = Optional.empty();
+        executor = Optional.empty();
+    }
+
+    @Override
+    public OutputSocket[] getOutputSockets() {
+        return new OutputSocket[]{frameOutputSocket, frameRateOutputSocket};
+    }
+
+    /**
+     * Starts the video capture from the
+     * @param deviceNumber The type of the device the webcam should be attached with
+     */
+    public synchronized void startVideo(final int deviceNumber) {
+        final OpenCVFrameConverter.ToMat convertToMat = new OpenCVFrameConverter.ToMat();
+        final OpenCVFrameGrabber grabber = new OpenCVFrameGrabber(deviceNumber);
+        if(this.executor.isPresent()){
+            throw new IllegalStateException("The video retrieval thread has already been started.");
+        }
+        if(this.grabber.isPresent()){
+            throw new IllegalStateException("The Frame Grabber has already been started.");
+        }
+        try {
+            grabber.start();
+        } catch (FrameGrabber.Exception e) {
+            throw new IllegalStateException("A problem occurred trying to start the frame grabber", e);
+        }
+        // Store the grabber only once it has been started in the case that there is an exception.
+        this.grabber = Optional.of(grabber);
+
+        final Thread frameExecutor = new Thread(() -> {
+            long lastFrame = System.currentTimeMillis();
+            while(!Thread.interrupted()) {
+                final Frame videoFrame;
+                try {
+                    videoFrame = grabber.grab();
+                } catch (FrameGrabber.Exception e) {
+                    throw new IllegalStateException("Failed to grab image", e);
+                }
+                frameOutputSocket.setValue(convertToMat.convert(videoFrame));
+                long thisMoment = System.currentTimeMillis();
+                frameRateOutputSocket.setValue(1000/(thisMoment - lastFrame));
+                lastFrame = thisMoment;
+            }
+        });
+        frameExecutor.setUncaughtExceptionHandler(
```

Comment:
Since the entire thread is defined above, it would probably be better to catch all exceptions within the thread.